README.md
# Resource Watch Carto Adapter

[![Build Status](https://travis-ci.com/resource-watch/rw-adapter-carto.svg?branch=dev)](https://travis-ci.com/resource-watch/rw-adapter-carto)
[![Test Coverage](https://api.codeclimate.com/v1/badges/0afec809bb5b7e1d37e7/test_coverage)](https://codeclimate.com/github/resource-watch/rw-adapter-carto/test_coverage)

This repository is the microservice that implements the carto adapter
funcionality, which is exposed on the /carto endpoint.

## Dependencies

You will need [Control Tower](https://github.com/control-tower/control-tower) up and running - either natively or with Docker. Refer to the project's README for information on how to set it up.

The Carto microservice is built using [Node.js](https://nodejs.org/en/), and can be executed either natively or using Docker, each of which has its own set of requirements.

Native execution requires:
- [Node.js](https://nodejs.org/en/)

Execution using Docker requires:
- [Docker](https://www.docker.com/)
- [Docker Compose](https://docs.docker.com/compose/)

Dependencies on other Microservices:

- [Dataset](https://github.com/resource-watch/dataset/)
- [Converter](https://github.com/resource-watch/converter)

## Getting started

Start by cloning the repository from github to your execution environment

```
git clone https://github.com/resource-watch/rw-adapter-carto.git && cd rw-adapter-carto
```

After that, follow one of the instructions below:

### Using native execution

1 - Set up your environment variables. See `dev.env.sample` for a list of variables you should set, which are described in detail in [this section](#configuration-environment-variables) of the documentation. Native execution will NOT load the `dev.env` file content, so you need to use another way to define those values

2 - Install node dependencies using yarn:
```
yarn
```

3 - Start the application server:
```
yarn start
```

The endpoints provided by this microservice should now be available through Control Tower's URL.

### Using Docker

1 - Create and complete your `dev.env` file with your configuration. The meaning of the variables is available in this [section](#configuration-environment-variables). You can find an example `dev.env.sample` file in the project root.

2 - Execute the following command to run Control tower:

```
./carto.sh develop
```

The endpoints provided by this microservice should now be available through Control Tower's URL.

## Testing

There are two ways to run the included tests:

### Using native execution

Follow the instruction above for setting up the runtime environment for native execution, then run:
```
yarn test
```

### Using Docker

Follow the instruction above for setting up the runtime environment for Docker execution, then run:
```
./carto.sh test
```

## Configuration

### Environment variables

- PORT => TCP port in which the service will run
- NODE_PATH => relative path to the source code. Should be `app/src`
- CT_REGISTER_MODE => if `auto` the microservice automatically registers on Control Tower on start
- CT_TOKEN => 
- API_VERSION => API version identifier that prefixes the URL. Should be `v1`


CHANGELOG.md

## 10/03/2022

- Upgrade node 12 -> 16.
## 28/05/2021

- Update `rw-api-microservice-node` to add CORS support.

## 30/03/2021

- Modify API HTTP verbs to match public API method signature.

## 10/12/2020

- Replace CT integration library

# 1.1.0

## 17/11/2020

- Remove dependency on Control Tower's filter functionality.
- Security update of NPM package `dot-prop` from 4.2.0 to 4.2.1.
- Security update of NPM package `lodash` from 4.17.15 to 4.17.20.

# 1.0.0

## 09/04/2020

- Add node affinity to kubernetes configuration.
- Update node version to v12.x
- Replace NPM with Yarn
- Add healthcheck endpoint


app/test/e2e/src/mock.js
// eslint-disable-next-line import/no-extraneous-dependencies
const nock = require('nock');
const { DEFAULT_RESPONSE_SQL_QUERY } = require('./test-constants');

const createMockConvertSQL = sqlQuery => nock(process.env.CT_URL, { encodedQueryParams: true })
    .get(`/v1/convert/sql2SQL?sql=${encodeURIComponent(sqlQuery)}&experimental=true&raster=false`)
    .reply(200, {
        data: {
            type: 'result',
            id: 'undefined',
            attributes: {
                query: 'SELECT * FROM test',
                jsonSql: { select: [{ value: '*', alias: null, type: 'literal' }], from: 'test' }
            },
            relationships: {}
        }
    });

const createMockSQLQuery = (sqlQuery, connectorUrl) => nock(connectorUrl)
    .get(`/api/v2/sql?q=select%20*%20from%20wdpa_protected_areas%20limit%200`)
    .reply(200, DEFAULT_RESPONSE_SQL_QUERY);

const createMockSQLQueryPOST = sqlQuery => nock('https://wri-01.carto.com')
    .post('/api/v2/sql', { q: sqlQuery })
    .reply(200, DEFAULT_RESPONSE_SQL_QUERY);

const createMockSQLCount = () => nock(`https://wri-01.carto.com`)
    .post('/api/v2/sql')
    .reply(200, {
        rows: [{
            count: 2
        }]
    });

const createMockRegisterDataset = id => nock(process.env.CT_URL)
    .patch(`/v1/dataset/${id}`)
    .reply(200, {});

module.exports = {
    createMockConvertSQL,
    createMockSQLQuery,
    createMockSQLCount,
    createMockSQLQueryPOST,
    createMockRegisterDataset
};


app/test/e2e/carto.spec.js
const nock = require('nock');
const chai = require('chai');
const { getTestServer } = require('./src/test-server');

const should = chai.should();

const requester = getTestServer();

const dataset = {
    data: {
        id: '00c47f6d-13e6-4a45-8690-897bdaa2c723',
        attributes: {
            connectorUrl: 'https://wri-01.carto.com/tables/wdpa_protected_areas/table',
            table_name: 'wdpa_protected_areas'
        }
    }
};

const fields = [{
    field1: {
        type: 'number'
    },
    the_geom: {
        type: 'geometry'
    }
}];

describe('E2E test', () => {

    before(async () => {
        nock.cleanAll();

        if (process.env.NODE_ENV !== 'test') {
            throw Error(`Running the test suite with NODE_ENV ${process.env.NODE_ENV} may result in permanent data loss. Please use NODE_ENV=test.`);
        }
    });

    it('Get fields correctly', async () => {
        nock(`https://wri-01.carto.com`)
            .get(encodeURI(`/api/v2/sql?q=select * from ${dataset.data.attributes.table_name} limit 0`))
            .reply(200, {
                rows: [],
                fields
            });


        const response = await requester
            .post(`/api/v1/carto/fields/${dataset.data.id}`)
            .send({
                dataset,
                loggedUser: null
            });

        response.status.should.equal(200);
        response.body.should.have.property('tableName');
        response.body.tableName.should.equal(dataset.data.attributes.table_name);
        response.body.should.have.property('fields');
        response.body.fields.should.deep.equal(fields);
    });

    afterEach(() => {
        if (!nock.isDone()) {
            throw new Error(`Not all nock interceptors were used: ${nock.pendingMocks()}`);
        }
    });
});


app/test/e2e/query.spec.js
const nock = require('nock');
const chai = require('chai');
const { getTestServer } = require('./utils/test-server');
const { ensureCorrectError, createMockGetDataset } = require('./utils/helpers');
const { createMockConvertSQL, createMockSQLCount, createMockSQLQueryPOST } = require('./utils/mock');
const { DEFAULT_RESPONSE_SQL_QUERY } = require('./utils/test-constants');

chai.should();

const requester = getTestServer();

nock.disableNetConnect();
nock.enableNetConnect(process.env.HOST_IP);

describe('Query tests', () => {
    before(async () => {
        nock.cleanAll();

        if (process.env.NODE_ENV !== 'test') {
            throw Error(`Running the test suite with NODE_ENV ${process.env.NODE_ENV} may result in permanent data loss. Please use NODE_ENV=test.`);
        }
    });

    it('Query to dataset without connectorType cartodb should fail', async () => {
        const timestamp = new Date().getTime();

        createMockGetDataset(timestamp, { connectorType: 'foo' });

        const requestBody = {};

        const query = `select * from ${timestamp}`;

        const response = await requester
            .post(`/api/v1/carto/query/${timestamp}?sql=${encodeURI(query)}`)
            .send(requestBody);

        response.status.should.equal(422);
        response.body.should.have.property('errors').and.be.an('array').and.have.lengthOf(1);
        response.body.errors[0].detail.should.include('This operation is only supported for datasets with connectorType \'rest\'');
    });

    it('Query to dataset without a supported provider should fail', async () => {
        const timestamp = new Date().getTime();

        createMockGetDataset(timestamp, { provider: 'foo' });

        const requestBody = {};

        const query = `select * from ${timestamp}`;

        const response = await requester
            .post(`/api/v1/carto/query/${timestamp}?sql=${encodeURI(query)}`)
            .send(requestBody);

        response.status.should.equal(422);
        response.body.should.have.property('errors').and.be.an('array').and.have.lengthOf(1);
        response.body.errors[0].detail.should.include('This operation is only supported for datasets with provider \'cartodb\'');
    });

    it('Query without sql or fs parameter should return bad request', async () => {
        const timestamp = new Date().getTime();

        createMockGetDataset(timestamp);

        const response = await requester
            .post(`/api/v1/carto/query/${timestamp}`)
            .send();

        ensureCorrectError(response, 'sql or fs required', 400);
    });

    it('Send query should return result(happy case)', async () => {
        const timestamp = new Date().getTime();
        const sql = 'SELECT * FROM test LIMIT 2 OFFSET 0';

        createMockGetDataset(timestamp);

        createMockSQLCount();
        createMockSQLQueryPOST(sql);
        createMockConvertSQL(sql);

        const response = await requester
            .post(`/api/v1/carto/query/${timestamp}`)
            .query({ sql })
            .send();


        response.status.should.equal(200);
        response.body.should.have.property('data').and.instanceOf(Array);
        response.body.should.have.property('meta').and.instanceOf(Object);

        const { meta, data } = response.body;
        data.should.deep.equal(DEFAULT_RESPONSE_SQL_QUERY.rows);

        meta.should.have.property('cloneUrl').and.instanceOf(Object);
        // eslint-disable-next-line camelcase
        const { cloneUrl: { http_method, url, body } } = meta;
        http_method.should.equal('POST');
        url.should.equal(`/dataset/${timestamp}/clone`);
        body.should.have.property('dataset').and.instanceOf(Object);

        const { datasetUrl, application } = body.dataset;
        application.should.deep.equal(['your', 'apps']);
        datasetUrl.should.equal(`/query/${timestamp}?sql=${encodeURI(sql).replace('*', '%2A')}`);
    });

    afterEach(() => {
        if (!nock.isDone()) {
            throw new Error(`Not all nock interceptors were used: ${nock.pendingMocks()}`);
        }
    });
});


app/src/app.js
const Koa = require('koa');
const logger = require('logger');
const koaLogger = require('koa-logger');
const config = require('config');
const loader = require('loader');
const convert = require('koa-convert');
const koaSimpleHealthCheck = require('koa-simple-healthcheck');
const ErrorSerializer = require('serializers/error.serializer');
const { RWAPIMicroservice } = require('rw-api-microservice-node');

const koaBody = require('koa-body')({
    multipart: true,
    jsonLimit: '50mb',
    formLimit: '50mb',
    textLimit: '50mb'
});


const app = new Koa();


app.use(convert(koaBody));

app.use(async (ctx, next) => {
    try {
        await next();
    } catch (err) {
        let error = err;
        try {
            error = JSON.parse(err);
        } catch (e) {
            logger.error(`Error parsing exception: ${err.message}`);
        }
        ctx.status = err.status || 500;
        if (ctx.status >= 500) {
            logger.error(err);
        } else {
            logger.info(err);
        }
        ctx.body = ErrorSerializer.serializeError(ctx.status, error.message);
        if (process.env.NODE_ENV === 'prod' && this.status === 500) {
            ctx.body = 'Unexpected error';
        }
        ctx.response.type = 'application/vnd.api+json';
    }

});

app.use(koaLogger());
app.use(koaSimpleHealthCheck());

app.use(RWAPIMicroservice.bootstrap({
    name: config.get('service.name'),
    info: require('../microservice/register.json'),
    swagger: require('../microservice/public-swagger.json'),
    logger,
    baseURL: process.env.CT_URL,
    url: process.env.LOCAL_URL,
    token: process.env.CT_TOKEN,
    fastlyEnabled: process.env.FASTLY_ENABLED,
    fastlyServiceId: process.env.FASTLY_SERVICEID,
    fastlyAPIKey: process.env.FASTLY_APIKEY
}));

loader.loadRoutes(app);


const instance = app.listen(process.env.PORT, () => {
    if (process.env.CT_REGISTER_MODE === 'auto') {
        RWAPIMicroservice.register().then(() => {
            logger.info('CT registration process started');
        }, (error) => {
            logger.error(error);
            process.exit(1);
        });
    }
});
logger.info('Server started in ', process.env.PORT);


module.exports = instance;


app/Gruntfile.js
module.exports = function (grunt) {

    grunt.file.setBase('..');
    require('load-grunt-tasks')(grunt);

    grunt.initConfig({

        clean: {},
        express: {
            dev: {
                options: {
                    script: 'app/index.js',
                    node_env: 'dev',
                    port: process.env.PORT,
                    output: 'started'
                }
            }
        },

        mochaTest: {
            unit: {
                options: {
                    reporter: 'spec',
                    quiet: false, // Optionally suppress output to standard out (defaults to false)
                    clearRequireCache: true, // Optionally clear the require cache before running tests (defaults to false)
                },
                src: ['app/test/unit/**/*.test.js']
            },
            e2e: {
                options: {
                    reporter: 'spec',
                    quiet: false, // Optionally suppress output to standard out (defaults to false)
                    clearRequireCache: true, // Optionally clear the require cache before running tests (defaults to false)
                },
                src: ['app/test/e2e/**/*.spec.js']
            }
        },
        watch: {
            options: {
                livereload: 35730
            },
            jssrc: {
                files: [
                    'app/src/**/*.js',
                ],
                tasks: ['mochaTest:unit', 'express:dev'],
                options: {
                    spawn: false
                }
            },
            unitTest: {
                files: [
                    'app/test/unit/**/*.test.js',
                ],
                tasks: ['mochaTest:unit'],
                options: {
                    spawn: false
                }
            },
            e2eTest: {
                files: [
                    'app/test/e2e/**/*.spec.js',
                ],
                tasks: ['mochaTest:e2e'],
                options: {
                    spawn: true
                }
            },

        }
    });


    grunt.registerTask('unitTest', ['mochaTest:unit']);

    grunt.registerTask('e2eTest', ['mochaTest:e2e']);

    grunt.registerTask('e2eTest-watch', ['watch:e2eTest']);

    grunt.registerTask('test', ['unitTest']);

    grunt.registerTask('serve', ['express:dev', 'watch']);

    grunt.registerTask('default', 'serve');

};


app/src/services/carto.service.js
const logger = require('logger');
const url = require('url');
const request = require('request');
const requestPromise = require('request-promise');

const Json2sql = require('sql2json').json2sql;
const Sql2json = require('sql2json').sql2json;

class CartoService {

    static async getFields(urlDataset, tableName) {
        logger.debug(`Obtaining fields of ${url} and table ${tableName}`);
        const parsedUrl = url.parse(urlDataset);
        logger.debug('Doing request to ', `https://${parsedUrl.host}/api/v2/sql?q=select * from ${tableName} limit 0`);
        try {
            const result = await requestPromise({
                method: 'GET',
                uri: `https://${parsedUrl.host}/api/v2/sql?q=select * from ${tableName} limit 0`,
                json: true
            });
            return result.fields;
        } catch (err) {
            logger.error('Error obtaining fields', err);
            throw new Error('Error obtaining fields');
        }
    }

    static async getCount(urlDataset, tableName, where) {
        logger.debug(`Obtaining count of ${urlDataset} and table ${tableName}`);
        const parsed = new Sql2json(`select count(*) from ${tableName}`).toJSON();

        if (where) {
            parsed.where = where;
        }
        const sql = Json2sql.toSQL(parsed);
        const parsedUrl = url.parse(urlDataset);
        logger.debug('Doing request to ', unescape(`https://${parsedUrl.host}/api/v2/sql?q=${sql}`));
        try {
            const result = await requestPromise({
                method: 'POST',
                uri: `https://${parsedUrl.host}/api/v2/sql`,
                json: true,
                body: {
                    q: unescape(sql)
                }
            });
            return result.rows[0].count;
        } catch (err) {
            logger.error('Error obtaining count', err.message);
            throw err;
        }
    }

    static executeQuery(urlDataset, query, format) {
        logger.debug(`Doing query ${query} to ${urlDataset}`);
        const parsedUrl = url.parse(urlDataset);
        logger.debug('Doing request to ', `https://${parsedUrl.host}/api/v2/sql?q=${query}`);
        const configRequest = {
            uri: `https://${parsedUrl.host}/api/v2/sql`,
            method: 'POST',
            json: true,
            body: {
                q: unescape(query)
            }
        };
        if (format && format === 'geojson') {
            configRequest.body.format = 'geojson';
        }
        return request(configRequest);
    }

}

module.exports = CartoService;


app/src/services/query.service.js
const logger = require('logger');
const CartoService = require('services/carto.service');
const JSONStream = require('JSONStream');
const json2csv = require('json2csv');
const Json2sql = require('sql2json').json2sql;

class QueryService {

    constructor(sql, dataset, passthrough, cloneUrl, download, downloadType, jsonSql) {
        this.sql = sql;
        this.dataset = dataset;
        this.passthrough = passthrough;
        this.cloneUrl = cloneUrl;
        this.download = download;
        this.downloadType = downloadType;
        this.jsonSql = jsonSql;
        this.pagination = 100;
        this.timeout = false;
        this.timeoutFunc = setTimeout(() => {
            this.timeout = true;
        }, 60000);
    }

    async init() {
        await this.getCount();
    }

    async getCount() {
        logger.debug('Obtaining count', this.jsonSql);
        this.count = await CartoService.getCount(this.dataset.connectorUrl, this.jsonSql.from, this.jsonSql.where);
        if (this.jsonSql.limit && this.count > this.jsonSql.limit) {
            this.count = this.jsonSql.limit;
        }
        if ((this.jsonSql.limit && this.jsonSql.limit > this.pagination) || !this.jsonSql.limit) {
            this.jsonSql.limit = this.pagination;
        }
        logger.debug('LIMIT ', this.jsonSql.limit, ' count', this.count);
    }

    convertObject(data) {
        if (this.download && this.downloadType === 'csv') {
            return `${json2csv({
                data,
                hasCSVColumnTitle: this.first
            })}\n`;
        }
        return `${!this.first ? ',' : ''}${JSON.stringify(data)}`;

    }

    async writeRequest(request, format) {
        let count = 0;
        return new Promise((resolve, reject) => {
            let parser = null;
            if (format === 'geojson') {
                parser = JSONStream.parse('features.*');
            } else {
                parser = JSONStream.parse('rows.*');
            }
            request.pipe(parser)
                .on('data', (data) => {
                    logger.debug('data', data);
                    count++;
                    this.passthrough.write(this.convertObject(data));
                    this.first = false;
                })
                .on('end', () => resolve(count))
                .on('error', () => reject(new Error('Error in stream')));
        });
    }

    async execute() {
        logger.info('Executing query');
        const pages = Math.ceil(this.count / this.pagination);
        this.first = true;
        if (!this.download) {
            this.passthrough.write(`{"data":[`);
            if (this.downloadType === 'geojson') {
                this.passthrough.write(`{"type": "FeatureCollection", "features": [`);
            }
        } else if (this.download) {
            if (this.downloadType === 'geojson') {
                this.passthrough.write(`{"data":[{"type": "FeatureCollection", "features": [`);
            } else if (this.downloadType !== 'csv') {
                this.passthrough.write(`[`);
            }
        }

        for (let i = 0; i < pages; i++) {
            if (this.timeout) {
                break;
            }
            logger.debug(`Obtaining page ${i}`);
            const offset = i * this.pagination;
            if (i + 1 === pages) {
                this.jsonSql.limit = this.count - (this.pagination * i);
            }
            logger.debug('Query', `${Json2sql.toSQL(this.jsonSql)} OFFSET ${offset}`);
            const request = CartoService.executeQuery(this.dataset.connectorUrl, `${Json2sql.toSQL(this.jsonSql)} OFFSET ${offset}`, this.downloadType);

            const count = await this.writeRequest(request, this.downloadType);
            // if not return the same number of rows that pagination is that the query finished
            if (count < this.pagination) {
                break;
            }
        }

        if (this.timeout) {
            this.passthrough.end();
            throw new Error('Timeout exceeded');
        }
        clearTimeout(this.timeoutFunc);
        const meta = {
            cloneUrl: this.cloneUrl
        };

        if (!this.download) {

            if (this.downloadType === 'geojson') {
                this.passthrough.write(`]}`);
            }
            this.passthrough.write(`], "meta": ${JSON.stringify(meta)} }`);
        } else if (this.download) {
            if (this.downloadType === 'geojson') {
                this.passthrough.write(`]}]}`);
            } else if (this.downloadType !== 'csv') {
                this.passthrough.write(`]`);
            }
        }
        logger.debug('Finished');
        this.passthrough.end();
    }


}

module.exports = QueryService;


app/src/middleware/dataset.middleware.js
const logger = require('logger');
const DatasetService = require('services/dataset.service');

class DatasetMiddleware {

    static async getDatasetById(ctx, next) {
        const datasetId = ctx.params.dataset;
        logger.debug('[DatasetRouter - getDatasetById] - Dataset id', datasetId);

        if (!datasetId) {
            ctx.throw(400, 'Invalid request');
        }

        const dataset = await DatasetService.getDatasetById(datasetId);

        if (!dataset) {
            ctx.throw(404, 'Dataset not found');
        }

        if (dataset.attributes.connectorType !== 'rest') {
            ctx.throw(422, 'This operation is only supported for datasets with connectorType \'rest\'');
        }

        if (dataset.attributes.provider !== 'cartodb') {
            ctx.throw(422, 'This operation is only supported for datasets with provider \'cartodb\'');
        }

        ctx.request.body.dataset = {
            id: dataset.id,
            ...dataset.attributes
        };

        await next();
    }

}

module.exports = DatasetMiddleware;


app/src/routes/api/v1/carto.router.js
const Router = require('koa-router');
const logger = require('logger');
const { RWAPIMicroservice } = require('rw-api-microservice-node');
const CartoService = require('services/carto.service');
const QueryService = require('services/query.service');
const FieldSerializer = require('serializers/field.serializer');
const passThrough = require('stream').PassThrough;
const ErrorSerializer = require('serializers/error.serializer');
const DatasetMiddleware = require('middleware/dataset.middleware');

const router = new Router({
    prefix: '/carto',
});

const serializeObjToQuery = (obj) => Object.keys(obj).reduce((a, k) => {
    a.push(`${k}=${encodeURIComponent(obj[k])}`);
    return a;
}, []).join('&');


class CartoRouter {

    static getCloneUrl(url, idDataset) {
        return {
            http_method: 'POST',
            url: `/dataset/${idDataset}/clone`,
            body: {
                dataset: {
                    datasetUrl: url.replace('/carto', ''),
                    application: ['your', 'apps']
                }
            }
        };
    }

    static async query(ctx) {
        ctx.set('Content-type', 'application/json');
        const { format } = ctx.query;
        const cloneUrl = CartoRouter.getCloneUrl(ctx.request.url, ctx.params.dataset);
        try {
            ctx.body = passThrough();
            const queryService = await new QueryService(ctx.query.sql, ctx.request.body.dataset, ctx.body, cloneUrl, false, format, ctx.state.jsonSql);
            await queryService.init();
            queryService.execute();
            logger.debug('Finished query');
        } catch (err) {
            logger.error(err);
            ctx.body = ErrorSerializer.serializeError(err.statusCode || 500, err.error && err.error.error ? err.error.error[0] : err.message);
            ctx.status = 500;
        }
    }

    static async download(ctx) {
        try {
            ctx.body = passThrough();
            const format = ctx.query.format ? ctx.query.format : 'csv';
            let mimetype;
            switch (format) {

                case 'csv':
                    mimetype = 'text/csv';
                    break;
                case 'json':
                default:
                    mimetype = 'application/json';
                    break;

            }

            const cloneUrl = CartoRouter.getCloneUrl(ctx.request.url, ctx.params.dataset);
            const queryService = await new QueryService(ctx.query.sql, ctx.request.body.dataset, ctx.body, cloneUrl, true, format, ctx.state.jsonSql);
            await queryService.init();
            ctx.set('Content-disposition', `attachment; filename=${ctx.request.body.dataset.id}.${format}`);
            ctx.set('Content-type', mimetype);

            queryService.execute();
            logger.debug('Finished query');
        } catch (err) {
            ctx.body = ErrorSerializer.serializeError(err.statusCode || 500, err.error && err.error.error ? err.error.error[0] : err.message);
            ctx.status = 500;
        }
    }

    static async fields(ctx) {
        logger.info('Obtaining fields');
        const fields = await CartoService.getFields(ctx.request.body.dataset.connectorUrl, ctx.request.body.dataset.tableName);
        ctx.body = FieldSerializer.serialize(fields, ctx.request.body.dataset.tableName);
    }

    static async registerDataset(ctx) {
        logger.info('Registering dataset with data', ctx.request.body);
        try {
            await CartoService.getFields(ctx.request.body.connector.connectorUrl, ctx.request.body.connector.tableName);
            await RWAPIMicroservice.requestToMicroservice({
                method: 'PATCH',
                uri: `/dataset/${ctx.request.body.connector.id}`,
                body: {
                    dataset: {
                        status: 1
                    }
                },
                json: true
            });
        } catch (e) {
            await RWAPIMicroservice.requestToMicroservice({
                method: 'PATCH',
                uri: `/dataset/${ctx.request.body.connector.id}`,
                body: {
                    dataset: {
                        status: 2,
                        errorMessage: `${e.name} - ${e.message}`
                    }
                },
                json: true
            });
        }
        ctx.body = {};
    }

    static async deleteDataset(ctx) {
        logger.info('Delete carto dataset stub');
        ctx.body = {};
    }

}

const sanitizeUrl = async (ctx, next) => {
    if (ctx.request.body.dataset && ctx.request.body.dataset.connectorUrl && /\/u\//.test(ctx.request.body.dataset.connectorUrl)) {
        const user = ctx.request.body.dataset.connectorUrl.split(/\/u|\/table/)[1].replace('/', '');
        ctx.request.body.dataset.connectorUrl = `https://${user}.carto.com/api/v2/sql`;
    } else if (ctx.request.body.connector && ctx.request.body.connector.connector_url && /\/u\//.test(ctx.request.body.connector.connector_url)) {
        const user = ctx.request.body.connector.connector_url.split(/\/u|\/table/)[1].replace('/', '');
        ctx.request.body.connector.connector_url = `https://${user}.carto.com/api/v2/sql`;
    }
    await next();
};


const toSQLMiddleware = async (ctx, next) => {
    const options = {
        method: 'GET',
        json: true,
        resolveWithFullResponse: true,
        simple: false
    };
    if (!ctx.query.sql && !ctx.request.body.sql && !ctx.query.outFields && !ctx.query.outStatistics) {
        ctx.throw(400, 'sql or fs required');
        return;
    }

    if (ctx.query.sql || ctx.request.body.sql) {
        logger.debug('Checking sql correct');
        const params = { ...ctx.query, ...ctx.request.body };
        options.uri = `/convert/sql2SQL?sql=${encodeURIComponent(params.sql)}&experimental=true&raster=${ctx.request.body.dataset.type === 'raster'}`;
        if (params.geostore) {
            options.uri += `&geostore=${params.geostore}`;
        }
        if (params.geojson) {
            options.body = {
                geojson: params.geojson
            };
            options.method = 'POST';
        }
    } else {
        logger.debug('Obtaining sql from featureService');
        const fs = { ...ctx.request.body };
        delete fs.dataset;
        const query = serializeObjToQuery(ctx.request.query);
        const body = fs;
        const resultQuery = { ...query };

        if (resultQuery) {
            options.uri = `/convert/fs2SQL${resultQuery}&tableName=${ctx.request.body.dataset.tableName}`;
        } else {
            options.uri = `/convert/fs2SQL?tableName=${ctx.request.body.dataset.tableName}`;
        }
        options.body = body;
        options.method = 'POST';
    }

    try {
        const result = await RWAPIMicroservice.requestToMicroservice(options);
        logger.debug('result', result.statusCode);
        if (result.statusCode === 204 || result.statusCode === 200) {
            ctx.query.sql = result.body.data.attributes.query;
            ctx.state.jsonSql = result.body.data.attributes.jsonSql;
            await next();
        } else if (result.statusCode === 400) {
            ctx.status = result.statusCode;
            ctx.body = result.body;
        } else {
            ctx.throw(result.statusCode, result.body);
        }

    } catch (e) {
        if (e.errors && e.errors.length > 0 && e.errors[0].status >= 400 && e.errors[0].status < 500) {
            ctx.status = e.errors[0].status;
            ctx.body = e;
        } else {
            throw e;
        }
    }
};

router.get('/query/:dataset', DatasetMiddleware.getDatasetById, toSQLMiddleware, sanitizeUrl, CartoRouter.query);
router.post('/query/:dataset', DatasetMiddleware.getDatasetById, toSQLMiddleware, sanitizeUrl, CartoRouter.query);
router.get('/download/:dataset', DatasetMiddleware.getDatasetById, toSQLMiddleware, sanitizeUrl, CartoRouter.download);
router.post('/download/:dataset', DatasetMiddleware.getDatasetById, toSQLMiddleware, sanitizeUrl, CartoRouter.download);
router.get('/fields/:dataset', DatasetMiddleware.getDatasetById, sanitizeUrl, CartoRouter.fields);
router.post('/rest-datasets/cartodb', sanitizeUrl, CartoRouter.registerDataset);
router.delete('/rest-datasets/cartodb/:dataset', CartoRouter.deleteDataset);

module.exports = router;
